server/balancer.go

```diff
@@ -468,183 +475,268 @@ func (h *balanceHotRegionScheduler) calculateScore(cluster *clusterInfo) {
+			}
+		}
 	}
 }
 
-func (h *balanceHotRegionScheduler) SelectSourceRegion(cluster *clusterInfo) *RegionInfo {
+func (h *balanceHotRegionScheduler) balanceByPeer(cluster *clusterInfo) (*RegionInfo, *metapb.Peer, *metapb.Peer) {
 	var (
-		maxWritten             uint64
-		sourceStore            uint64
+		maxWrittenBytes        uint64
+		srcStoreID             uint64
 		maxHotStoreRegionCount int
 	)
-	// choose a hot store as transfer source
-	// the numbers of the hot regions in that store has higher priority than TotalWrittenBytes
-	for sid, s := range h.scoreStatus {
-		if s.RegionsStat.Len() < 2 {
+
+	// get the srcStoreId
+	for storeID, statistics := range h.statisticsAsPeer {
+		if statistics.RegionsStat.Len() < 2 {
 			continue
 		}
 
-		if maxHotStoreRegionCount < s.RegionsStat.Len() {
-			maxHotStoreRegionCount = s.RegionsStat.Len()
-			maxWritten = s.TotalWrittenBytes
-			sourceStore = sid
+		if maxHotStoreRegionCount < statistics.RegionsStat.Len() {
+			maxHotStoreRegionCount = statistics.RegionsStat.Len()
+			maxWrittenBytes = statistics.WrittenBytes
+			srcStoreID = storeID
 			continue
 		}
 
-		if maxHotStoreRegionCount == s.RegionsStat.Len() && maxWritten < s.TotalWrittenBytes {
-			maxWritten = s.TotalWrittenBytes
-			sourceStore = sid
+		if maxHotStoreRegionCount == statistics.RegionsStat.Len() && maxWrittenBytes < statistics.WrittenBytes {
+			maxWrittenBytes = statistics.WrittenBytes
+			srcStoreID = storeID
 		}
 	}
-
-	if sourceStore == 0 {
-		return nil
+	if srcStoreID == 0 {
+		return nil, nil, nil
 	}
 
-	length := h.scoreStatus[sourceStore].RegionsStat.Len()
-	// the hottest region in the store not move
-	// randomly pick a region from 1 .. length-1
-	// TODO: consider hot degree when pick
-	rr := h.r.Intn(length-1) + 1
-	pickedRegionStat := h.scoreStatus[sourceStore].RegionsStat[rr]
-	if pickedRegionStat.antiCount < hotRegionAntiCount {
-		return nil
+	stores := cluster.getStores()
+	var destStoreID uint64
+	for _, i := range h.r.Perm(h.statisticsAsPeer[srcStoreID].RegionsStat.Len()) {
+		rs := h.statisticsAsPeer[srcStoreID].RegionsStat[i]
+		srcRegion := cluster.getRegion(rs.RegionID)
+		if len(srcRegion.DownPeers) != 0 || len(srcRegion.PendingPeers) != 0 {
+			continue
+		}
+
+		var filters []Filter
+		filters = append(filters, newExcludedFilter(srcRegion.GetStoreIds(), srcRegion.GetStoreIds()))
+		filters = append(filters, newDistinctScoreFilter(h.opt.GetReplication(), stores, cluster.getLeaderStore(srcRegion)))
+		filters = append(filters, newStateFilter(h.opt))
+		filters = append(filters, newStorageThresholdFilter(h.opt))
+		destStoreIDs := make([]uint64, 0, len(stores))
+		for _, store := range stores {
+			if filterTarget(store, filters) {
+				continue
+			}
+			destStoreIDs = append(destStoreIDs, store.GetId())
+		}
+
+		destStoreID = h.selectDestStoreByPeer(destStoreIDs, srcRegion, srcStoreID)
+		if destStoreID != 0 {
+			srcRegion.WrittenBytes = rs.WrittenBytes
+			h.adjustBalanceLimitByPeer(srcStoreID)
```

Comment:
If the peer is relatively balanced, the limit will not adjust when balance leader.